<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


you are running the lab in your  browser, so we will install the libraries using `piplite`


In [1]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [12]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

<class 'AttributeError'>: module 'pexpect' has no attribute 'TIMEOUT'

In [3]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser


In [2]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.

</p>


In [8]:
path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

you will need to download the dataset; if you are running locally, please comment out the following


In [10]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path,"auto.csv")
path="auto.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [13]:
# Import pandas library
#import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"

df = pd.read_csv(path,header=None,error_bad_lines=False)

<ipython-input-13-e0a20b41e153>:6: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  df = pd.read_csv(path,header=None,error_bad_lines=False)
Skipping line 207: expected 26 fields, saw 29
Skipping line 208: expected 26 fields, saw 29
Skipping line 209: expected 26 fields, saw 29
Skipping line 210: expected 26 fields, saw 29
Skipping line 211: expected 26 fields, saw 29
Skipping line 212: expected 26 fields, saw 29
Skipping line 213: expected 26 fields, saw 29
Skipping line 214: expected 26 fields, saw 29
Skipping line 215: expected 26 fields, saw 29
Skipping line 216: expected 26 fields, saw 29
Skipping line 217: expected 26 fields, saw 29
Skipping line 218: expected 26 fields, saw 29
Skipping line 219: expected 26 fields, saw 29
Skipping line 220: expected 26 fields, saw 29
Skipping line 221: expected 26 fields, saw 29
Skipping line 222: expected 26 fields, saw 29
Skipping line 223: expected 26 

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [21]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


0    1            2    3    4     5            6    7      8     9   ...  \
0  3.0    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1  3.0    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2  1.0    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3  2.0  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4  2.0  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

      16    17    18    19    20   21    22    23    24     25  
0  130.0  mpfi  3.47  2.68   9.0  111  5000  21.0  27.0  13495  
1  130.0  mpfi  3.47  2.68   9.0  111  5000  21.0  27.0  16500  
2  152.0  mpfi  2.68  3.47   9.0  154  5000  19.0  26.0  16500  
3  109.0  mpfi  3.19  3.40  10.0  102  5500  24.0  30.0  13950  
4  136.0  mpfi  3.19  3.40   8.0  115  5500  18.0  22.0  17450  

[5 rows x 26 columns]

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [22]:
# Write your code below and press Shift+Enter to execute 
print('The last 10 rows of the dataframe')
df.tail(10)

The last 10 rows of the dataframe


0                   1      2       3      4     5      6     7   \
196 -2.000000                 103  volvo     gas    std  four  sedan   rwd   
197 -1.000000                  74  volvo     gas    std  four  wagon   rwd   
198 -2.000000                 103  volvo     gas  turbo  four  sedan   rwd   
199 -1.000000                  74  volvo     gas  turbo  four  wagon   rwd   
200 -1.000000                  95  volvo     gas    std  four  sedan   rwd   
201 -1.000000                  95  volvo     gas  turbo  four  sedan   rwd   
202 -1.000000                  95  volvo     gas    std  four  sedan   rwd   
203 -1.000000                  95  volvo  diesel  turbo  four  sedan   rwd   
204 -1.000000                  95  volvo     gas  turbo  four  sedan   rwd   
205  0.826526  0.9083333333333334   54.3    2385   ohcf  four    108  2bbl   

        8       9   ...          16    17    18    19    20   21    22    23  \
196  front  104.30  ...  141.000000  mpfi  3.78  3.15   9.5  114  5400  24.0   
197  front  104.30  ...  141.000000  mpfi  3.78  3.15   9.5  114  5400  24.0   
198  front  104.30  ...  130.000000  mpfi  3.62  3.15   7.5  162  5100  17.0   
199  front  104.30  ...  130.000000  mpfi  3.62  3.15   7.5  162  5100  17.0   
200  front  109.10  ...  141.000000  mpfi  3.78  3.15   9.5  114  5400  23.0   
201  front  109.10  ...  141.000000  mpfi  3.78  3.15   8.7  160  5300  19.0   
202  front  109.10  ...  173.000000  mpfi  3.58  2.87   8.8  134  5500  18.0   
203  front  109.10  ...  145.000000   idi  3.01  3.40  23.0  106  4800  26.0   
204  front  109.10  ...  141.000000  mpfi  3.78  3.15   9.5  114  5400  19.0   
205   3.62    2.64  ...    9.791667   Low     0     1   NaN  NaN   NaN   NaN   

       24     25  
196  28.0  15985  
197  28.0  16515  
198  22.0  18420  
199  22.0  18950  
200  28.0  16845  
201  25.0  19045  
202  23.0  21485  
203  27.0  22470  
204  25.0  22625  
205   NaN    NaN  

[10 rows x 26 columns]

<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe our data, we can introduce a header. This information is available at:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Thus, we have to add headers manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [23]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


We replace headers and recheck our dataframe:


In [24]:
df.columns = headers
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0        3.0                 ?  alfa-romero       gas        std          two   
1        3.0                 ?  alfa-romero       gas        std          two   
2        1.0                 ?  alfa-romero       gas        std          two   
3        2.0               164         audi       gas        std         four   
4        2.0               164         audi       gas        std         four   
5        2.0                 ?         audi       gas        std          two   
6        1.0               158         audi       gas        std         four   
7        1.0                 ?         audi       gas        std         four   
8        1.0               158         audi       gas      turbo         four   
9        0.0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...        130.0   
1  convertible          rwd           front        88.6  ...        130.0   
2    hatchback          rwd           front        94.5  ...        152.0   
3        sedan          fwd           front        99.8  ...        109.0   
4        sedan          4wd           front        99.4  ...        136.0   
5        sedan          fwd           front        99.8  ...        136.0   
6        sedan          fwd           front       105.8  ...        136.0   
7        wagon          fwd           front       105.8  ...        136.0   
8        sedan          fwd           front       105.8  ...        131.0   
9    hatchback          4wd           front        99.5  ...        131.0   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000     21.0   
1         mpfi  3.47    2.68               9.0        111      5000     21.0   
2         mpfi  2.68    3.47               9.0        154      5000     19.0   
3         mpfi  3.19    3.40              10.0        102      5500     24.0   
4         mpfi  3.19    3.40               8.0        115      5500     18.0   
5         mpfi  3.19    3.40               8.5        110      5500     19.0   
6         mpfi  3.19    3.40               8.5        110      5500     19.0   
7         mpfi  3.19    3.40               8.5        110      5500     19.0   
8         mpfi  3.13    3.40               8.3        140      5500     17.0   
9         mpfi  3.13    3.40               7.0        160      5500     16.0   

  highway-mpg  price  
0        27.0  13495  
1        27.0  16500  
2        26.0  16500  
3        30.0  13950  
4        22.0  17450  
5        25.0  15250  
6        25.0  17710  
7        25.0  18920  
8        20.0  23875  
9        22.0      ?  

[10 rows x 26 columns]

We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:


In [25]:
df1 = df.replace('?',np.NaN)
df1.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0        3.0               NaN  alfa-romero       gas        std          two   
1        3.0               NaN  alfa-romero       gas        std          two   
2        1.0               NaN  alfa-romero       gas        std          two   
3        2.0               164         audi       gas        std         four   
4        2.0               164         audi       gas        std         four   
5        2.0               NaN         audi       gas        std          two   
6        1.0               158         audi       gas        std         four   
7        1.0               NaN         audi       gas        std         four   
8        1.0               158         audi       gas      turbo         four   
9        0.0               NaN         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...        130.0   
1  convertible          rwd           front        88.6  ...        130.0   
2    hatchback          rwd           front        94.5  ...        152.0   
3        sedan          fwd           front        99.8  ...        109.0   
4        sedan          4wd           front        99.4  ...        136.0   
5        sedan          fwd           front        99.8  ...        136.0   
6        sedan          fwd           front       105.8  ...        136.0   
7        wagon          fwd           front       105.8  ...        136.0   
8        sedan          fwd           front       105.8  ...        131.0   
9    hatchback          4wd           front        99.5  ...        131.0   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000     21.0   
1         mpfi  3.47    2.68               9.0        111      5000     21.0   
2         mpfi  2.68    3.47               9.0        154      5000     19.0   
3         mpfi  3.19    3.40              10.0        102      5500     24.0   
4         mpfi  3.19    3.40               8.0        115      5500     18.0   
5         mpfi  3.19    3.40               8.5        110      5500     19.0   
6         mpfi  3.19    3.40               8.5        110      5500     19.0   
7         mpfi  3.19    3.40               8.5        110      5500     19.0   
8         mpfi  3.13    3.40               8.3        140      5500     17.0   
9         mpfi  3.13    3.40               7.0        160      5500     16.0   

  highway-mpg  price  
0        27.0  13495  
1        27.0  16500  
2        26.0  16500  
3        30.0  13950  
4        22.0  17450  
5        25.0  15250  
6        25.0  17710  
7        25.0  18920  
8        20.0  23875  
9        22.0    NaN  

[10 rows x 26 columns]

We can drop missing values along the column "price" as follows:


In [26]:
df = df1.dropna(subset=['price'],axis=0)
df.head(10)

symboling normalized-losses         make fuel-type aspiration  \
0         3.0               NaN  alfa-romero       gas        std   
1         3.0               NaN  alfa-romero       gas        std   
2         1.0               NaN  alfa-romero       gas        std   
3         2.0               164         audi       gas        std   
4         2.0               164         audi       gas        std   
5         2.0               NaN         audi       gas        std   
6         1.0               158         audi       gas        std   
7         1.0               NaN         audi       gas        std   
8         1.0               158         audi       gas      turbo   
10        2.0               192          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0         130.0         mpfi  3.47    2.68               9.0        111   
1         130.0         mpfi  3.47    2.68               9.0        111   
2         152.0         mpfi  2.68    3.47               9.0        154   
3         109.0         mpfi  3.19    3.40              10.0        102   
4         136.0         mpfi  3.19    3.40               8.0        115   
5         136.0         mpfi  3.19    3.40               8.5        110   
6         136.0         mpfi  3.19    3.40               8.5        110   
7         136.0         mpfi  3.19    3.40               8.5        110   
8         131.0         mpfi  3.13    3.40               8.3        140   
10        108.0         mpfi  3.50    2.80               8.8        101   

    peak-rpm city-mpg highway-mpg  price  
0       5000     21.0        27.0  13495  
1       5000     21.0        27.0  16500  
2       5000     19.0        26.0  16500  
3       5500     24.0        30.0  13950  
4       5500     18.0        22.0  17450  
5       5500     19.0        25.0  15250  
6       5500     19.0        25.0  17710  
7       5500     19.0        25.0  18920  
8       5500     17.0        20.0  23875  
10      5800     23.0        29.0  16430  

[10 rows x 26 columns]

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [27]:
# Write your code below and press Shift+Enter to execute 
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


In [28]:
df.to_csv('automobile.csv',index=False)

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [30]:
df.dtypes

symboling            float64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size          float64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg             float64
highway-mpg          float64
price                 object
dtype: object

A series with the data type of each column is returned.


In [31]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

symboling            float64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size          float64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg             float64
highway-mpg          float64
price                 object
dtype: object


<p>
As shown above, it is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [33]:
df.describe(include='all')

symboling normalized-losses    make fuel-type aspiration  \
count   201.000000               164     201       201        201   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       181        165   
mean      0.840796               NaN     NaN       NaN        NaN   
std       1.254802               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           199        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    201.000000          201   197     197        201.000000        199   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           92    23      19               NaN         19   
mean     126.875622          NaN   NaN     NaN         10.164279        NaN   
std       41.546834          NaN   NaN     NaN          4.004965        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       98.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        199  201.000000  201.000000   201  
unique        22         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          36         NaN         NaN     2  
mean         NaN   25.179104   30.686567   NaN  
std          NaN    6.423220    6.815150   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [ ]:
# describe all the columns in "df" 
df.describe(include = "all")

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [35]:
# Write your code below and press Shift+Enter to execute 
df[['length','compression-ratio']].describe()

length  compression-ratio
count  201.000000         201.000000
mean   174.200995          10.164279
std     12.322175           4.004965
min    141.100000           7.000000
25%    166.800000           8.600000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

<details><summary>Click here for the solution</summary>

```python
df[['length', 'compression-ratio']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    float64
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
